Add guidelines for node PR triage (#5436)
kubernetes/community: contributors/devel/sig-node/triage.md

# SIG Node Triage Process

Currently, SIG Node is limiting its triage review to pull requests. In the
near-future, we hope to expand to issues as well.

This is intended to guide developers through the SIG Node triage process. You
will learn about the roles of authors, reviewers, and approvers, as well as
what labels are required for each stage of the PR process.

We track most node PRs on the [SIG Node Triage project board]. All test and
CI-related PRs are tracked on the [CI subproject board].

[SIG Node Triage project board]: https://github.com/orgs/kubernetes/projects/49

> [CI subproject board]: https://github.com/orgs/kubernetes/projects/43

Reviewer: wanna add a link to the handy bot commands too? - https://go.k8s.io/bot-commands

## Triage

All new PRs added to the board should begin in the **Triage** column.

When a pull request is made against kubernetes/kubernetes, it will typically

> have the following [labels]:

Reviewer 2: could you make explicit the prow add them by default when the PR is opened?

- needs-triage
- needs-priority
- needs-ok-to-test (not needed for project members)

In order to be moved out of this column, all of the above labels must be set.

[labels]: https://github.com/kubernetes/test-infra/blob/master/label_sync/labels.md

### needs-ok-to-test

Use your best judgment in determining whether a PR is OK to test. You don't
have to do a full review: just make sure that the code does not appear to be
actively malicious, and the PR appears to be doing something useful. Use the
command `/ok-to-test`.

If the PR is trivial and doesn't provide much value, feel free to close it and
link the author to the [trivial edits policy].

[trivial edits policy]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#trivial-edits

### needs-priority

You can take a quick look at what the PR is addressing and then apply a
priority label.

> - **priority/critical-urgent:** Urgent bug fix, required ASAP.

Reviewer: `critical-urgen` also tells the release team that they can't make a release.

Reviewer 3: if it's critical-urgent feels like maybe we should do something more than just tag it? maybe also post in sig-node channel to flag?

- **priority/important-soon:** Needs to be completed this release. Important
  bug fixes + KEPs targeted for the current milestone.
- **priority/important-longterm:** Has an attached issue/KEP, but unclear what
  the specific priority is.
- **priority/backlog:** Non-urgent changes such as typo fixes, linters, small
  comment fixes, etc.

### needs-triage

There are two more things to check before accepting a PR for triage.

The first is whether the SIG is correct. If the PR does not appear to touch SIG
Node code or require a SIG Node approver, you should remove the SIG Node label,
and add other SIG labels as appropriate.

The second is verifying the kind of PR. Most will be bug fixes, cleanups, or
documentation. Feature PRs should generally have an attached KEP. API changes
and deprecations require special review; those labels may be mistakenly
applied, so check over the PR to make sure they're accurate.

Once you've quickly looked over a PR, applied the appropriate labels, and it
looks ready to proceed to review (i.e. it doesn't have any labels that would
mean it's waiting on more work from the author), you can mark the PR as triaged

> with `/triage accepted`.

Reviewer 2: who can `/triage accepted`? It can be worth clarifying the roles suitable to do the procedure.

Reviewer: today we don't have special roles in sig-node other than the sig-chair and TL. So for now i'd say any regular contributor to sig-node should be able to do this.

Author: Technically anyone who's a Kubernetes org member can apply the label. I've definitely done some drive-by triage for other SIGs :)

## Waiting on Author

This column means that the PR is waiting on some action from the author. A
reviewer may have requested changes, or a PR may have one of the following
do-not-merge labels:

- **do-not-merge/hold:** usually set by a reviewer
- **do-not-merge/work-in-progress:** usually set by an author

> - **do-not-merge/release-note-needed:** needs a release note, override with

Reviewer 3: Note: there's something fussy that happens when there _is_ a release note but the bot doesn't pick it up and the author has to try to fix the text in their description - i've recently seen authors try to update their notes and bot still doesn't pick it up (also can't be manually overridden to add `release-note`).

Reviewer 3: I guess what Im saying is that if it only has dnm/release-note-needed it's prob not sufficient to stay in this column?

Author: If we can remove the label by setting `/release-note-none`, as long as there's no other issues it can move. If not, it requires action from the author to fix the release note. I'll try to make that explicit.

  `/release-note-none`
- **do-not-merge/contains-merge-commits:** PR needs to be rebased
- **needs-rebase:** PR needs to be rebased

PRs that do not have any of the above labeled in this column should be
evaluated occasionally to see if they are ready for review.

## Waiting on Reviewer

This PR needs review! If you're not sure how to review a PR, start by
familiarizing yourself with the Kubernetes [pull request guidelines] and
[review guidelines].

PRs in this column must have the following labels set:

- triage-accepted
- priority
- kind

TODO: Add some node-specific stuff here.

[pull request guidelines]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md
[review guidelines]: https://github.com/kubernetes/community/blob/master/contributors/guide/review-guidelines.md

## Waiting on Approver

These PRs are waiting on an `approved` label that can only be provided by

> approvers. The bot will always tell you whose approval is needed on what files.

Reviewer: files->directories.

CI-related PRs are tracked on the [CI subproject board].

For help with the commands listed in the document below, review the [bot command documentation].

[SIG Node Triage project board]: https://github.com/orgs/kubernetes/projects/49
[CI subproject board]: https://github.com/orgs/kubernetes/projects/43
[bot command documentation]: https://go.k8s.io/bot-commands

## Triage

All new PRs added to the board should begin in the **Triage** column.

When a pull request is made against kubernetes/kubernetes, it will typically
have the following [labels], applied by the Prow bot:

- needs-triage
- needs-priority
- needs-ok-to-test (not needed for project members)

In order to be moved out of this column, all of the above labels must be set.

[labels]: https://github.com/kubernetes/test-infra/blob/master/label_sync/labels.md

### needs-ok-to-test

Use your best judgment in determining whether a PR is OK to test. You don't
have to do a full review: just make sure that the code does not appear to be
actively malicious, and the PR appears to be doing something useful. Use the
command `/ok-to-test`.

Only [Kubernetes org members] can add this label.

> If the PR is trivial and doesn't provide much value, feel free to close it and

Reviewer 4: ```suggestion
If the PR is trivial and doesn't provide much value, feel free to close it (using `/close` command) and
```

> using the `/close` command and link the author to the [trivial edits policy].
> 
> [Kubernetes org members]: https://github.com/kubernetes/community/blob/master/community-membership.md#member
> [trivial edits policy]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#trivial-edits

Reviewer 4: ```suggestion
using the `/close` command and link the author to the [trivial edits policy].

[Kubernetes org members]: https://github.com/kubernetes/community/blob/master/community-membership.md#member
[trivial edits policy]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#trivial-edits

>  `/close`
>
> `Thank you for your PR. It seems to only contain trivial edits. Please read our [trivial edits policy](https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#trivial-edits) and reopen with the description if you believe that this PR is needed.`
```

### needs-priority

You can take a quick look at what the PR is addressing and then apply a
priority label.

Anyone can add this label.

- **priority/critical-urgent:** Urgent bug fix, required ASAP. If not
  addressed, will block a release. These issues should always be discussed in
  the `#sig-node` channel on Slack.
- **priority/important-soon:** Needs to be completed this release. Important
  bug fixes + KEPs targeted for the current milestone.
- **priority/important-longterm:** Has an attached issue/KEP, but unclear what
  the specific priority is.

> - **priority/backlog:** Non-urgent changes such as typo fixes, linters, small

Reviewer 4: Maybe have backlog before important-longeterm to demonstrate the order of priorities

Reviewer 4: according to https://github.com/kubernetes/test-infra/blob/master/label_sync/labels.md

Author: These should be in order already? critical-urgent > important-soon > important-longterm > backlog > awaiting-more-evidence (which I didn't include)

Reviewer 4: My understanding before was that backlog is higher priority to look at as its something that can be looked at immediately. Important longterm may not even be clear on design. But I agree with this ordering, thank you for clarification!

> - **priority/backlog:** Non-urgent changes such as typo fixes, linters, small
>   comment fixes, etc.

Reviewer 4: Also I'd go with other examples:

```suggestion
- **priority/backlog:** Non-urgent changes such as minor performance optimizations, improving error logs, increasing tests code coverage, code refactoring, and addressing code analysis issues.
```

### needs-triage

There are two more things to check before accepting a PR for triage.

The first is whether the SIG is correct. If the PR does not appear to touch SIG
Node code or require a SIG Node approver, you should remove the SIG Node label,
and add other SIG labels as appropriate.

The second is verifying the kind of PR. Most will be bug fixes, cleanups, or
documentation. Feature PRs should generally have an attached KEP. API changes
and deprecations require special review; those labels may be mistakenly
applied, so check over the PR to make sure they're accurate.

> Once you've quickly looked over a PR, applied the appropriate labels, and it

Reviewer 4: Also mention to check the release note accuracy. It is often forgotten

Author: I will add that to the review section; I don't think it's necessary as part of triage.

looks ready to proceed to review (i.e. it doesn't have any labels that would
mean it's waiting on more work from the author), you can mark the PR as triaged
with `/triage accepted`.

Only [Kubernetes org members] can add this label.

## Waiting on Author

This column means that the PR is waiting on some action from the author. A

> reviewer may have requested changes, or a PR may have one of the following

Reviewer 4: ```suggestion
reviewer may have requested changes, tests might be failing, or a PR may have one of the following
```

Author: I typically do *not* put PRs into **Waiting on Author** if tests are failing because tests are rather flaky, only if there is a confirmed test failure that needs to be addressed (and usually at that point I put a hold on it).

Reviewer 4: Maybe we should to save reviewers time. It's easier to review when you know that tests are passing. Maybe link instruction on how to investigate test failures.

do-not-merge labels:

- **do-not-merge/hold:** usually set by a reviewer
- **do-not-merge/work-in-progress:** usually set by an author
- **do-not-merge/release-note-needed:** needs a release note to be added by the
  author, [Kubernetes org members] can override with `/release-note-none` if
  not required
- **do-not-merge/contains-merge-commits:** PR needs to be rebased
- **needs-rebase:** PR needs to be rebased

> PRs that do not have any of the above labeled in this column should be
> evaluated occasionally to see if they are ready for review.

Reviewer 4: ```suggestion
Author of the PR is encouraged to resolve or reply to all the comments and comment on PR to indicate that is ready for the next iteration of reviews.

PRs that do not have any of the above labeled in this column will be
evaluated occasionally to see if they are ready for review.

If left untouched for 30 days, PR will be marked as rotten and will eventually be closed.
```

## Waiting on Reviewer

This PR needs review! If you're not sure how to review a PR, start by
familiarizing yourself with the Kubernetes [pull request guidelines] and
[review guidelines].

PRs in this column must have the following labels set:

- triage-accepted
- priority
- kind

Only [Kubernetes org members] can add an `/lgtm`.

If you want to become an official Node reviewer, you should read through the
[reviewer responsibilities and requirements].

> TODO: Add some node-specific stuff here.

Reviewer 4: what do you envision to have here?

Author: More node specific content on reviewing, code structure, etc. for components we control would be nice.

Author: (Not in the scope of this PR so I'm going to close this)

[pull request guidelines]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md
[review guidelines]: https://github.com/kubernetes/community/blob/master/contributors/guide/review-guidelines.md
[reviewer responsibilities and requirements]: https://github.com/kubernetes/community/blob/master/community-membership.md#reviewer

## Waiting on Approver

These PRs are waiting on an `approved` label that can only be provided by
approvers. The bot will always tell you whose approval is needed on which
directories.

PRs in this column must have the following labels set:

> - lgtm

Reviewer 4: ```suggestion
- lgtm (during triage some lgtm's from people not directly involved with SIG Node may not be enough for approver to look at the PR)
```

Author: I'll add a note about that outside of this list.

- triage-accepted
- priority
- kind

> Check for the bot's comment to see which files still need approvers from the

Reviewer 4: can you copy/paste an example here? Might be helpful

appropriate OWNERS. If the PR already has an approval for the node components
(commonly, anything in `./pkg/kubelet/*`), you can mark the PR as Done manually
while waiting on other approvers.

> Anyone can `/approve` a PR, but the bot will only pay attention to the

Reviewer 4: I wouldn't encourage anyone to approve. It creates noise in PR comments.

Author: I don't disagree but people do it :smile: I don't think this is encouraging it, but I can rewrite.

approvals from [Kubernetes approvers].

[Kubernetes approvers]: https://github.com/kubernetes/community/blob/master/community-membership.md#approver

## Done

> This column is for closed and merged PRs. Huzzah!

Reviewer 4: will this column contain PRs that are waiting for approval from other OWNERS, API review, or release managers for cherry-picks?
